jobbergate-cli/jobbergate_cli/subapps/applications/tools.py

```diff
@@ -171,53 +161,138 @@ def upload_application(
+    Abort.require_condition(module_file_path.is_file(), f"Application module file {module_file_path} does not exist")
+
+    application_config = load_application_config_from_source(config_file_path.read_text())
+
+    logger.debug("Preparing to upload the template configuration")
+
+    response_code = cast(
+        int,
+        make_request(
+            jg_ctx.client,
+            f"/jobbergate/job-script-templates/{application_id}",
+            "PUT",
+            expect_response=False,
+            abort_message="Request to upload application configuration was not accepted by the API",
+            support=True,
+            json={"template_vars": application_config.application_config},
+        ),
+    )
+
+    if response_code != 200:
+        return False
+
+    supporting_files = application_config.jobbergate_config.supporting_files or []
+
+    for complete_template_path in itertools.chain(application_path.rglob("*.j2"), application_path.rglob("*.jinja2")):
+        relative_template_path = complete_template_path.relative_to(application_path)
+        logger.debug(f"Preparing to upload {relative_template_path}")
+
+        with open(complete_template_path, "r", newline="") as template_file:
+            if relative_template_path.as_posix() in supporting_files:
+                file_type = FileType.SUPPORT
+            else:
+                file_type = FileType.ENTRYPOINT
+
+            response_code = cast(
+                int,
+                make_request(
+                    jg_ctx.client,
+                    f"/jobbergate/job-script-templates/{application_id}/upload/template/{file_type}",
+                    "PUT",
+                    expect_response=False,
+                    abort_message="Request to upload application files was not accepted by the API",
+                    support=True,
+                    files={"upload_file": (relative_template_path.as_posix(), template_file, "text/plain")},
+                ),
+            )
+            if response_code != 200:
+                return False
+
+    logger.debug(f"Preparing to upload {JOBBERGATE_APPLICATION_MODULE_FILE_NAME}")
+    with open(module_file_path, "r", newline="") as module_file:
         response_code = cast(
             int,
             make_request(
                 jg_ctx.client,
-                f"/jobbergate/applications/{application_id}/upload",
-                "POST",
+                f"/jobbergate/job-script-templates/{application_id}/upload/workflow",
+                "PUT",
                 expect_response=False,
-                abort_message="Request to upload application files was not accepted by the API",
+                abort_message="Request to upload application module was not accepted by the API",
                 support=True,
-                files=upload_files,
+                files={"upload_file": (module_file_path.name, module_file, "text/plain")},
+                json={"data": application_config.jobbergate_config},
```

Comment:
Note to myself to review this request.

I'm having a hard time packing the file and the json data in the same multi-part form. I asserted this work from the API swagger, so the back-end is ok.

Reply:
solved in [b56a1e7](https://github.com/omnivector-solutions/jobbergate/pull/263/commits/b56a1e7cd447ba5bcdfd4195e85e06eaf311118b)